### PR TITLE
plugins/asdf: Add `asdf` extendable version manager plugin

### PR DESCRIPTION
**Pull request by RobLoach**
This adds a plugin to load the [asdf version manager](https://github.com/asdf-vm/asdf) through *oh-my-bash*. It's roughly based off of the [original pull request](https://github.com/ohmybash/oh-my-bash/pull/310), moves to the bash syntax, and allows checking for the `ASDF_DIR` variable before hitting the file system.

```
asdf plugin add nodejs
asdf install nodejs latest
```

### CODE REVIEW

File: completions/asdf.completion.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#! bash oh-my-bash.module
+# Bash completion support for the `asdf` command.
+# Depends on the `asdf` plugin.
+
+# Only load the completions if the ASDF_DIR variable was set.
+if [[ -v ASDF_DIR ]]; then
```
Comment (akinomyoga) [inline completions/asdf.completion.sh:6]: Oh My Bash targets Bash >= 3.2 while `[[ -v ASDF_DIR ]]` is only available in Bash >= 4.2.

```suggestion
if [[ ${ASDF_DIR+set} ]]; then
```

File: plugins/asdf/asdf.plugin.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#! bash oh-my-bash.module
+# asdf.plugin.sh: asdf plugin for oh-my-bash.
+
+# Custom ASDF_DIR location
+if [[ -v ASDF_DIR ]]; then
```
Comment (akinomyoga) [inline plugins/asdf/asdf.plugin.sh:5]: Similar to the previous comment.

File: plugins/asdf/asdf.plugin.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  . "${ASDF_DIR}/asdf.sh"
+
+# Arch Linux / AUR Package
+elif [[ -f "/opt/asdf-vm/asdf.sh" ]]; then
+  ASDF_DIR="/opt/asdf-vm"
+  . /opt/asdf-vm/asdf.sh
+
+# Homebrew
+elif _omb_util_command_exists brew; then
+  __ASDF_PREFIX="$(brew --prefix asdf)"
```
Comment (akinomyoga) [inline plugins/asdf/asdf.plugin.sh:20]: Is this an internal variable of this plugin? Can we change the variable name as follows?

```suggestion
  _omb_plugin_asdf__prefix="$(brew --prefix asdf)"
```

If so, the other occurrences of `__ASDF_PREFIX` should also be updated.

File: plugins/asdf/asdf.plugin.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  if [[ -f "$__ASDF_PREFIX/libexec/asdf.sh" ]]; then
+    ASDF_DIR="$__ASDF_PREFIX/libexec"
+    . "$ASDF_DIR/asdf.sh"
+  fi
+  unset __ASDF_PREFIX
```
Comment (akinomyoga) [inline plugins/asdf/asdf.plugin.sh:25]: We want to explicitly specify `-v` to remove the variable (but not a function of the same name).

```suggestion
  unset -v __ASDF_PREFIX
```

File: plugins/asdf/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# asdf plugin
+
+Add [oh-my-bash](https://ohmybash.github.io) integration with [asdf](https://github.com/asdf-vm/asdf), the extendable version manager, with support for Ruby, Node.js, Elixir, Erlang and more.
+
+### Installation
```
Comment (akinomyoga) [inline plugins/asdf/README.md:5]: I guess it must be heading level 2. The same for `### Usage`.

```suggestion
## Installation
```

**Review by akinomyoga — CHANGES_REQUESTED** [review]
> It's roughly based off of the [original pull request](https://github.com/ohmybash/oh-my-bash/pull/310), moves to the bash syntax, and allows checking for the `ASDF_DIR` variable before hitting the file system.

Thank you. Could you mention the original author (https://github.com/prodrigues1912) and URL of  PR #310 in the code comment around the top of `asdf.plugin.sh`? Since it will not remain in the git history, we need to leave the information manually.

**Comment by RobLoach** [comment]
Thanks a lot for the review, I'll fix these up.

File: plugins/asdf/asdf.plugin.sh (hunk 2 above)
Comment (RobLoach) [reply plugins/asdf/asdf.plugin.sh:5]:
``` suggestion
if [[ -v ${ASDF_DIR+set} ]]; then
```

File: plugins/asdf/README.md (hunk 5 above)
Comment (akinomyoga) [reply plugins/asdf/README.md]: > The same for `### Usage`.

File: plugins/asdf/asdf.plugin.sh (hunk 3 above)
Comment (akinomyoga) [reply plugins/asdf/asdf.plugin.sh]: > If so, the other occurrences of `__ASDF_PREFIX` should also be updated.

File: plugins/asdf/asdf.plugin.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#! bash oh-my-bash.module
+# asdf.plugin.sh: asdf plugin for oh-my-bash.
```
Comment (RobLoach) [inline plugins/asdf/asdf.plugin.sh:2]:
``` suggestion
# asdf.plugin.sh: asdf plugin for oh-my-bash.
# Author: Rob Loach (https://robloach.net)
# Author: prodrigues1912 (https://github.com/prodrigues1912)
# Originally suggested in https://github.com/ohmybash/oh-my-bash/pull/310
# Fork of the oh-my-zsh asdf plugin
```

File: plugins/asdf/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    ``` sh
+    plugins=(asdf)
+    ```
+
+2. Enable the completions by adding it to your oh-my-bash `completions` definition in `~/.bashrc`.
+    ``` sh
+    completions=(asdf)
+    ```
+
+### Usage
```
Comment (RobLoach) [inline plugins/asdf/README.md:22]:
``` suggestion
## Usage
```

File: plugins/asdf/asdf.plugin.sh (hunk 3 above)
Comment (RobLoach) [reply plugins/asdf/asdf.plugin.sh]: Good catch, thanks. Fixed up.

File: plugins/asdf/asdf.plugin.sh (hunk 2 above)
Comment (RobLoach) [reply plugins/asdf/asdf.plugin.sh]: Good catch, thanks. Fixed.

Comment (akinomyoga) [reply plugins/asdf/asdf.plugin.sh]: Thanks

File: plugins/asdf/asdf.plugin.sh (hunk 3 above)
Comment (akinomyoga) [reply plugins/asdf/asdf.plugin.sh]: Thanks.

File: plugins/asdf/asdf.plugin.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+# Originally suggested in https://github.com/ohmybash/oh-my-bash/pull/310
+# Fork of the oh-my-zsh asdf plugin
+
+# Custom ASDF_DIR location
+if [[ ${ASDF_DIR+set} ]]; then
+  . "$ASDF_DIR/asdf.sh"
+
+# Standard install location
+elif [[ -f "${XDG_CONFIG_HOME:-$HOME}/.asdf/asdf.sh" ]]; then
+  ASDF_DIR="${XDG_CONFIG_HOME:-$HOME}/.asdf"
```
Comment (akinomyoga) [inline plugins/asdf/asdf.plugin.sh:14]: Is this location correct? `XDG_CONFIG_HOME` is defaulted to `$HOME/.config`. It's not normal to have a dotdir under `~/.config` (i.e. `~/.config/.asdf`). The global code search in GitHub only finds only your setting:

- [Code search results `/${XDG_CONFIG_HOME:-$HOME}\/.asdf/`](https://github.com/search?q=%2F%24%7BXDG_CONFIG_HOME%3A-%24HOME%7D%5C%2F.asdf%2F&amp;type=code)

The upstream discussions (https://github.com/asdf-vm/asdf/issues/687, https://github.com/asdf-vm/asdf/pull/1351) .only mention `$XDG_CONFIG_HOME/asdf`

**Review by akinomyoga — COMMENTED** [review]
Thank you for updating

File: plugins/asdf/asdf.plugin.sh (hunk 8 above)
Comment (RobLoach) [reply plugins/asdf/asdf.plugin.sh:14]: Perhaps something like this?
``` suggestion
# Home
elif [[ -f "$HOME/.asdf/asdf.sh" ]]; then
  ASDF_DIR="$HOME/.asdf"
  . "${ASDF_DIR}/asdf.sh"

# Config
elif [[ -f "$XDG_CONFIG_HOME/asdf/asdf.sh" ]]; then
  ASDF_DIR="$XDG_CONFIG_HOME/asdf"
```

Comment (akinomyoga) [reply plugins/asdf/asdf.plugin.sh:14]: That works, but please replace `$XDG_CONFIG_HOME` with `${XDG_CONFIG_HOME:-$HOME/.config}`.

**Review by akinomyoga — APPROVED** [review]

**Comment by akinomyoga** [comment]
Thank you.

**Comment by RobLoach** [comment]
Thank you so much for the reviews and merge :+1: ... Happy to have this in.

**Pull request merged**